keras_nlp/utils/preset_utils.py

```diff
@@ -155,6 +158,101 @@ def save_to_preset(
+):
+    config_path = get_file(preset, config_file)
+    if not os.path.exists(config_path):
+        if allow_incomplete:
+            logging.warning(
+                f"Tokenizer is missing from the preset directory {preset}."
+            )
+            return
+        else:
+            raise FileNotFoundError(
+                f"Tokenizer is missing from the preset directory {preset}."
+                "If you wish to upload model without a tokenizer, set `allow_incomplete=True`."
+            )
+    with open(config_path) as config_file:
+        config = json.load(config_file)
+    layer = keras.saving.deserialize_keras_object(config)
+
+    tokenizer = get_tokenizer(layer)
+    if tokenizer and config["assets"]:
+        for asset in config["assets"]:
+            asset_path = os.path.join(preset, asset)
+            if not os.path.exists(asset_path):
+                raise FileNotFoundError(
+                    f"Asset {asset} doesn't exist in the preset direcotry {preset}."
+                )
+        config_dir = os.path.dirname(config_path)
+        asset_dir = os.path.join(config_dir, TOKENIZER_ASSET_DIR)
+        tokenizer.load_assets(asset_dir)
+    else:
+        raise ValueError("Tokenizer or its asset are missing or invalid.")
+
+
+def _validate_backbone(preset, config_file="config.json"):
+    config_path = get_file(preset, config_file)
+    with open(config_path) as config_file:
+        config = json.load(config_file)
+    # Check if backbone is deserializable.
+    keras.saving.deserialize_keras_object(config)
+
+    if config["weights"]:
+        weights_path = os.path.join(preset, config["weights"])
+        if not os.path.exists(weights_path):
+            raise FileNotFoundError(
+                f"The weights file doesn't exist in preset directory {preset} ."
+            )
+    else:
+        raise ValueError("there is no wieghts config!")
+
+
+def _validate_files(preset, backbone_config_file, tokenizer_config_file):
+    # TODO: check if file sizes are reasonable.
+    # TODO: validate asset files.
+    backbone_config_path = get_file(preset, backbone_config_file)
+    with open(backbone_config_path) as config_file:
+        backbone_config = json.load(config_file)
+    valid_files = [
+        backbone_config_file,
+        tokenizer_config_file,
+        "metadata.json",
+        "assets",
+        backbone_config["weights"],
+    ]
+    files = os.listdir(preset)
+    for file in files:
+        if file not in valid_files:
+            raise ValueError(f"File {file} is an unexpected file.")
+
+
+@keras_nlp_export("keras_nlp.upload_preset")
+def upload_preset(
+    uri,
+    preset,
+    config_file="config.json",
+    tokenizer_config_file="tokenizer.json",
+    allow_incomplete=False,
+):
+    # Check if preset directory exists.
+    if not os.path.exists(preset):
+        raise FileNotFoundError(f"The preset directory {preset} doesn't exist.")
+
+    if uri.startswith(KAGGLE_PREFIX):
```

Comment:
I'm not quite sure if I fully understand this comment. I think it's good to separate validation from upload so I moved the validation code before this `if`. Right now, HF can do the following:
```
if uri.startswith(HF_PREFIX):
    hf_handle = uri.removeprefix(HF_PREFIX)
    hf_hub.model_upload(hf_handle, preset)
```
Do you mean to put this in a separate function to prevent repetition for different hubs?

Reply:
Yeah, let's scratch this comment. I was thinking to encapsulate the file download/upload methods to one spot in this file so it's easy to extend to and separated from our saving loading business logic. But I don't think there is a huge difference. Let's leave as is.

Reply:
> ```py
> if uri.startswith(HF_PREFIX):
>    hf_handle = uri.removeprefix(HF_PREFIX)
>    huggingface_hub.upload_folder(hf_handle, preset)
>```

Wonderful if I can add HF support with only this code change! :+1:

Reply:
Yeah, I think that's pretty much it!
And we can do this now that the PR is merged!